### PR TITLE
[feat][proxy] Support proxy limit maximum connections per IP

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

Pulsar has the `brokerMaxConnectionsPerIp` configuration at the
broker, we can use it to limit the maximum connections per IP.

The original motivation and PR here: https://github.com/apache/pulsar/pull/10754

IMO, we can also apply it to pulsar-proxy, because when a large
number of proxy accesses under the same IP (maybe due to some wrong
operations) will cause the proxy to accept too much wrong traffic and
cause service unstable.

Discussion link: https://lists.apache.org/thread/co00zx8ns3ksbccxs82lfh624rx1sht6

### Modifications

- Reuse `ConnectionController` for proxy part.
- Move `ConnectionController` to `broker-common` module.
- Add proxy configuration `maxConcurrentInboundConnectionsPerIp`
- Send `ServerError.NotAllowedError` command to prevent client reconnecting.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by Jason918 — APPROVED** [review]
LGTM

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 1]
```diff
@@ -144,22 +146,29 @@ public ProxyConnection(ProxyService proxyService, DnsAddressResolverGroup dnsAdd
...
-            state = State.Closing;
-            ctx.close();
+        SocketAddress rmAddress = ctx.channel().remoteAddress();
+        ConnectionController.State state = connectionController.increaseConnection(rmAddress);
+        if (!state.equals(ConnectionController.State.OK)) {
+            ctx.channel().writeAndFlush(Commands.newError(-1, ServerError.NotAllowedError,
+                    state.equals(ConnectionController.State.REACH_MAX_CONNECTION)
+                            ? "Reached the maximum number of connections"
+                            : "Reached the maximum number of connections on address" + rmAddress));
+            ctx.channel().close();
```
Comment (nicoloboschi) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:163]: should we close it in the writeAndFlush callback ?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/limiter/ConnectionController.java [hunk 2]
```diff
@@ -60,11 +60,11 @@ class DefaultConnectionController implements ConnectionController {
         private final boolean maxConnectionsLimitEnabled;
         private final boolean maxConnectionsLimitPerIpEnabled;
 
-        public DefaultConnectionController(ServiceConfiguration configuration) {
-            this.maxConnections = configuration.getBrokerMaxConnections();
-            this.maxConnectionPerIp = configuration.getBrokerMaxConnectionsPerIp();
-            this.maxConnectionsLimitEnabled = configuration.getBrokerMaxConnections() > 0;
-            this.maxConnectionsLimitPerIpEnabled = configuration.getBrokerMaxConnectionsPerIp() > 0;
+        public DefaultConnectionController(int maximumConnections, int maximumConnectionsPerId) {
```
[the remark this answers is not in the document]
Comment (mattisonchao) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/limiter/ConnectionController.java:63]: Fixed.

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: Applied.

**Review by nicoloboschi — APPROVED** [review]
LGTM

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 3]
```diff
@@ -144,22 +146,29 @@ public ProxyConnection(ProxyService proxyService, DnsAddressResolverGroup dnsAdd
...
+                            : "Reached the maximum number of connections on address" + rmAddress))
+                            .addListener(result -> ctx.close());
             ProxyService.REJECTED_CONNECTIONS.inc();
         }
     }
 
     @Override
     public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
         super.channelUnregistered(ctx);
+        connectionController.decreaseConnection(ctx.channel().remoteAddress());
```
Comment (codelipenghui) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:171]: We should add tests to ensure the connection count of the same remote address will decrease after the connection is closed.

Comment (mattisonchao) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:171]: fixed, thanks!

**Review by codelipenghui — APPROVED** [review]

**Review by Shoothzj — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**